functions/access.js

```diff
@@ -0,0 +1,116 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require(".");
+const { checkAuth } = require("./auth");
+const { uuidv4 } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  const venue = await admin.collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const granted = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (!granted.exists || !granted.tokens) {
+    return false;
+  }
+
+  if (Object.keys(granted.tokens).includes(token)) {
+    // Record that the token was checked
+    if (!granted.tokens[token].usedAt) {
+      granted.tokens[token].usedAt = [];
+    }
+    granted.tokens[token].usedAt.push(new Date().getTime());
+    await venue.ref.collection("accessgranted").doc(uid).set(granted);
+    return true;
+  }
+
+  return false;
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  return await venue.ref.collection("access").doc(method).get();
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = getAccessDoc(venueId, "password");
+  if (!access.exists || !access.password) {
+    return false;
+  }
+  return passwordsMatch(access.password, password);
```

Comment:
Just a note: should we trim() passwords as well?

Reply:
Why would we want to `trim()` passwords? https://github.com/sparkletown/sparkle/pull/1016#discussion_r559295977

Reply:
To avoid accidental spaces in the beginning or the end of the password input.

Reply:
_nods_ I guess my thoughts on this are sort of similar to what I was saying in the other comment. It feels like the thing that would be better handled at the source of input (UI/data validation) than at the comparison function itself.